packages/api/src/car.js

```diff
@@ -116,52 +116,56 @@ export async function handleCarUpload (request, env, ctx, car, uploadType = 'Car
     structure = 'Complete'
   }
 
-  const [{ cid, pins }, backupKey] = await Promise.all([
-    addToCluster(car, env),
-    backup(car, rootCid, user._id, env, structure)
-  ])
+  const bucketKey = await uploadToBucket(env.s3Client, env.s3BucketName, car, rootCid, user._id, structure)
 
   const xName = headers.get('x-name')
   let name = xName && decodeURIComponent(xName)
   if (!name || typeof name !== 'string') {
     name = `Upload at ${new Date().toISOString()}`
   }
 
-  const normalizedCid = normalizeCid(cid)
+  const normalizedCid = normalizeCid(rootCid.toString())
   await env.db.createUpload({
     user: user._id,
     authKey: authToken?._id,
     contentCid: normalizedCid,
-    sourceCid: cid,
+    sourceCid: rootCid.toString(),
     name,
     type: uploadType,
-    backupUrls: backupKey
-      ? [`https://${env.s3BucketName}.s3.${env.s3BucketRegion}.amazonaws.com/${backupKey}`]
-      : [],
-    pins,
+    backupUrls: [`https://${env.s3BucketName}.s3.${env.s3BucketRegion}.amazonaws.com/${bucketKey}`],
+    pins: [],
     dagSize
   })
 
   /** @type {(() => Promise<any>)[]} */
-  const tasks = []
+  const tasks = [async () => {
+    let pins = []
+    try {
+      pins = await addToCluster(car, env)
+    } catch (err) {
+      env.log.warn(new Error('failed to upload to cluster', { cause: err }))
+      pins = await pinToCluster(rootCid.toString(), env)
+    }
 
-  // Retrieve current pin status and info about the nodes pinning the content.
-  // Keep querying Cluster until one of the nodes reports something other than
-  // Unpinned i.e. PinQueued or Pinning or Pinned.
-  if (!pins.some(p => PIN_OK_STATUS.includes(p.status))) {
-    tasks.push(waitAndUpdateOkPins.bind(
-      null,
-      normalizedCid,
-      env.cluster,
-      env.db)
-    )
-  }
+    await env.db.upsertPins(pins.map(p => ({
+      status: p.status,
+      contentCid: normalizedCid,
+      location: p.location
+    })))
+
+    // Retrieve current pin status and info about the nodes pinning the content.
+    // Keep querying Cluster until one of the nodes reports something other than
+    // Unpinned i.e. PinQueued or Pinning or Pinned.
+    if (!pins.some(p => PIN_OK_STATUS.includes(p.status))) {
+      await waitAndUpdateOkPins(normalizedCid, env.cluster, env.db)
+    }
+  }]
 
   if (ctx.waitUntil) {
     tasks.forEach(t => ctx.waitUntil(t()))
   }
 
-  return new JSONResponse({ cid })
+  return new JSONResponse({ cid: rootCid.toString() })
```

Comment:
minor: We don't seem to use the CID object form in this method. We could just have `carStat(car)` return a property called `cid` as the string form, so we don't have to call `rootCid.toString()` a bunch o times.

Reply:
I looked and we use `rootCid.code` immediately after calling `carStat` so I've left `carStat` alone but just assigned the stringified version to a local var and used that everywhere 😉.

Reply:
aye, that's better. Note that `uploadToBucket` just takes the cid object and calls toString on it again tho.